src/sql.rs

```diff
@@ -299,10 +303,39 @@ impl Sql {
         }
     }
 
-    /// Allocates a connection and calls given function with the connection.
+    /// Locks the write transactions mutex.
+    /// We do not make all transactions
+    /// [IMMEDIATE](https://www.sqlite.org/lang_transaction.html#deferred_immediate_and_exclusive_transactions)
+    /// for more parallelism -- at least read transactions can be made DEFERRED to run in parallel
+    /// w/o any drawbacks. But if we make write transactions DEFERRED also w/o any external locking,
+    /// then they are upgraded from read to write ones on the first write statement. This has some
+    /// drawbacks:
+    /// - If there are other write transactions, we block the thread and the db connection until
+    ///   upgraded. Also if some reader comes then, it has to get next, less used connection with a
+    ///   worse per-connection page cache.
+    /// - If a transaction is blocked for more than busy_timeout, it fails with SQLITE_BUSY.
+    /// - Configuring busy_timeout is not the best way to manage transaction timeouts, we would
+    ///   prefer it to be integrated with Rust/tokio asyncs. Moreover, SQLite implements waiting
+    ///   using sleeps.
+    /// - If upon a successful upgrade to a write transaction the db has been modified by another
+    ///   one, the transaction is rolled back and retried. It is an extra work in terms of
```

Comment:
I didn't such xperiments, and there is no info in documentation on that, but if i wrote SQLite, i'd retry on my own (or better add an option for that) because if i've already took a write lock and that took a time, it's strange to just return it back and fail. At least this way i guarantee that the transaction will be done eventually -- if user will do retries on its own, they could retry infinitely.